### PR TITLE
chore: upgrade zig to 0.13.0

### PR DESCRIPTION
**Pull request by paperclover**
### What does this PR do?

- Fixes https://github.com/oven-sh/bun/issues/10726
- Fixes https://github.com/oven-sh/bun/issues/10730
- Fixes https://github.com/oven-sh/bun/issues/11861

As there were many changes, I took this as a time to start some needed tech debt cleanup, particularly reworking `build.zig`.

- bindgen, headers obj, and misctools are no longer configured here. these tools had not been maintained and have not been used by anyone on the team as far as i am aware of. i left this code in, but later i intend to fully remove these and replace with a better solution.
- added `zig build check`, which run a compile without any output files. when not emitting LLVMIR, the semantic analysis pass is about 7x faster
- **added `bun check-zig`** (aka `zig build check-all`) which does the above on all platforms. since sema is single-threaded, each instance runs on a separate thread, so it is still very fast. I get about 12 seconds for a full check.

Big changes to be aware of

- `std.posix` is mostly posix only things. handfuls of functions here dropped windows support
  - `std.posix.O` does not exist on windows
- comptime mutable memory behavior has changed. it's never valid to have a runtime pointer to a `comptime var`. the solution for these is to copy the data into a `const` and take a pointer to that.
- `@fieldParentPtr` takes two arguments, and uses RLS to find the parent type. it also is looser on alignment, which means you may need `@alignCast` if the pointer is misaligned.

### How did you verify your code works?

image

image

### CODE REVIEW

File: packages/bun-inspector-protocol/src/protocol/v8/index.d.ts [hunk 1]
```diff
@@ -11604,7 +11604,7 @@ export namespace V8 {
      * @event `Page.domContentEventFired`
      */
     export type DomContentEventFiredEvent = {
-      timestamp: Network.MonotonicTime;
+      timestamp: Network.monotonicTime;
```
Comment (dylan-conway) [inline packages/bun-inspector-protocol/src/protocol/v8/index.d.ts]: I think this was changed when replacing `.Monotonic`

File: packages/bun-inspector-protocol/src/protocol/v8/index.d.ts [hunk 2]
```diff
@@ -11869,7 +11869,7 @@ export namespace V8 {
        */
       loaderId: Network.LoaderId;
       name: string;
-      timestamp: Network.MonotonicTime;
+      timestamp: Network.monotonicTime;
```
Comment (dylan-conway) [inline packages/bun-inspector-protocol/src/protocol/v8/index.d.ts]: Same here

File: packages/bun-inspector-protocol/src/protocol/v8/index.d.ts [hunk 3]
```diff
@@ -11901,7 +11901,7 @@ export namespace V8 {
      * @event `Page.loadEventFired`
      */
     export type LoadEventFiredEvent = {
-      timestamp: Network.MonotonicTime;
+      timestamp: Network.monotonicTime;
```
Comment (dylan-conway) [inline packages/bun-inspector-protocol/src/protocol/v8/index.d.ts]: And here

File: src/bun.js/node/node_fs.zig [hunk 4]
```diff
@@ -5624,7 +5626,6 @@ pub const NodeFS = struct {
         if (args.recursive) {
             std.fs.cwd().deleteTree(args.path.slice()) catch |err| {
                 const errno: bun.C.E = switch (err) {
-                    error.InvalidHandle => .BADF,
                     error.AccessDenied => .PERM,
```
Comment (dylan-conway) [inline src/bun.js/node/node_fs.zig:5929]: Looks like `error.InvalidUtf8` is wasi only now. We need to add a `error.InvalidWtf8` case

File: src/bun.js/node/node_fs.zig [hunk 5]
```diff
@@ -5671,7 +5672,7 @@ pub const NodeFS = struct {
             // TODO: switch to an implementation which does not use any "unreachable"
             std.fs.cwd().deleteTree(args.path.slice()) catch |err| {
                 const errno: E = switch (err) {
-                    error.InvalidHandle => .BADF,
+                    // error.InvalidHandle => .BADF,
                     error.AccessDenied => .PERM,
```
Comment (dylan-conway) [inline src/bun.js/node/node_fs.zig:5977]: `error.InvalidWtf8` here too

File: src/bun.js/node/node_fs.zig [hunk 6]
```diff
@@ -5708,13 +5709,13 @@ pub const NodeFS = struct {
...
             // empircally, it seems to return AccessDenied when the
             // file is actually a directory on macOS.
             if (args.recursive and
                 (er == error.IsDir or er == error.NotDir or er == error.AccessDenied))
             {
-                std.os.rmdirZ(dest) catch |err| {
+                std.posix.rmdirZ(dest) catch |err| {
                     if (args.force) {
                         return Maybe(Return.Rm).success;
                     }
```
Comment (dylan-conway) [inline src/bun.js/node/node_fs.zig:6023]: `error.InvalidWtf8` here too

Comment (dylan-conway) [reply src/bun.js/node/node_fs.zig:6023]: and on 5759

File: src/cli/add_completions.zig [hunk 7]
```diff
@@ -66,7 +66,9 @@ pub const index: Index = if (Environment.isDebug) Index.initFill(&.{"OOMWorkArou
                 break;
             }
         }
-        array.set(@as(FirstLetter, @enumFromInt(i)), &record);
+
+        const cloned = record;
+        array.set(@as(FirstLetter, @enumFromInt(i)), &cloned);
```
Comment (dylan-conway) [inline src/cli/add_completions.zig:71]: Ah this is new to me. It needs to be `const`?

File: src/deps/libuv.zig [hunk 8]
```diff
@@ -240,7 +241,7 @@ const _O_SHORT_LIVED = 0x1000;
 const _O_SEQUENTIAL = 0x0020;
 const _O_RANDOM = 0x0010;
 
-// These **do not** map to std.os.O!
+// These **do not** map to std.posix.O!
```
Comment (dylan-conway) [inline src/deps/libuv.zig]: or `bun.O`

File: src/open.zig [hunk 9]
```diff
@@ -139,8 +139,8 @@ pub const Editor = enum(u8) {
     pub fn byFallbackPathForEditor(editor: Editor, out: ?*[]const u8) bool {
         if (bin_path.get(editor)) |paths| {
             for (paths) |path| {
-                if (std.os.open(path, 0, 0)) |opened| {
-                    std.os.close(opened);
+                if (std.fs.cwd().openFile(path, .{ })) |opened| {
```
Comment (dylan-conway) [inline src/open.zig]: fmt

File: test/js/web/fetch/fetch_headers.test.js [hunk 10]
```diff
@@ -1,4 +1,4 @@
-import { describe, it, expect, beforeAll, afterAll } from "bun:test";
+timport { describe, it, expect, beforeAll, afterAll } from "bun:test";
```
Comment (dylan-conway) [inline test/js/web/fetch/fetch_headers.test.js]:
```suggestion
import { describe, it, expect, beforeAll, afterAll } from "bun:test";
```

File: build.zig [hunk 11]
```diff
@@ -1,303 +1,169 @@
...
+const CrossTarget = std.zig.CrossTarget;
+const OptimizeMode = std.builtin.OptimizeMode;
+const Module = Build.Module;
+const fs = std.fs;
+const Version = std.SemanticVersion;
+const Arch = std.Target.Cpu.Arch;
 
-/// Do not rename this constant. It is scanned by some scripts to determine which zig version to install.
-const recommended_zig_version = "0.12.0-dev.1828+225fe6ddb";
+const Wyhash11 = @import("./src/wyhash.zig").Wyhash11;
```
Comment (dylan-conway) [inline build.zig]:
```suggestion
```

**Review by dylan-conway — CHANGES_REQUESTED** [review]
A few comments. Main thing is adding `error.InvalidWtf8` in four places in `node_fs.zig`

File: src/cli/add_completions.zig (hunk 7 above)
Comment (paperclover) [reply src/cli/add_completions.zig:71]: `var record: [count][]const u8 = undefined;`

is comptime mutable memory

`const cloned = record;` makes it constant

File: src/bun.js/WebKit
Comment (dylan-conway) [inline src/bun.js/WebKit:1]: Intentional commit?

File: build.zig [hunk 12]
```diff
@@ -205,6 +215,14 @@ pub fn build(b: *Build) !void {
         step.dependOn(&b.addInstallFile(bun_obj.getEmittedBin(), "bun-zig.o").step);
     }
 
+    // zig build windows-shim
+    {
+        var step = b.step("windows-shim", "Build the Windows shim (bun_shim_impl.exe + bun_shim_debug.exe)");
```
Comment (dylan-conway) [inline build.zig:219]: nice

File: src/bun.js/WebKit
Comment (paperclover) [reply src/bun.js/WebKit:1]: no

**Comment by paperdave** [comment]
image
windows branch fails. i must have missed some standard library changes

**Comment by paperdave** [comment]
Known bugs now

- Windows has broken napi

**Review by dylan-conway — APPROVED** [review]
:tada:

**Pull request merged**